### PR TITLE
Kraken: realtime level

### PR DESCRIPTION
**Pull request by antoine-de**
Technical refactoring, nothing added.

Change the boolean disruption_active to an enum that has 3 values: `theoric`, `adapted`, `realtime`

needed for the realtime integration

### CODE REVIEW

File: source/kraken/tests/fill_disruption_from_chaos_tests.cpp [hunk 1]
```diff
@@ -181,13 +181,13 @@ BOOST_AUTO_TEST_CASE(add_impact_on_stop_area) {
     BOOST_REQUIRE_EQUAL(b.data->pt_data->vehicle_journeys.size(), 4);
     BOOST_REQUIRE_EQUAL(b.data->pt_data->journey_patterns.size(), 2);
     bool has_adapted_vj = false;
-    for(const auto* vj: b.data->pt_data->vehicle_journeys){
-        if(vj->is_adapted){
+    for (const auto* vj: b.data->pt_data->vehicle_journeys) {
+        if (vj->realtime_level == nt::RTLevel::Adapted) {
```
Comment (TeXitoi) [inline source/kraken/tests/fill_disruption_from_chaos_tests.cpp]: switch to tests the different possibilities (and what to do if RT?)

File: source/kraken/tests/fill_disruption_from_chaos_tests.cpp [hunk 2]
```diff
@@ -240,18 +240,18 @@ BOOST_AUTO_TEST_CASE(add_impact_and_update_on_stop_area) {
...
         bool has_adapted_vj = false;
-        for(const auto* vj: data.pt_data->vehicle_journeys){
-            if(vj->adapted_validity_pattern->days.none() && vj->validity_pattern->days.none()){
+        for (const auto* vj: data.pt_data->vehicle_journeys) {
+            if (vj->adapted_validity_pattern->days.none() && vj->validity_pattern->days.none()) {
                 //some vj don't circulate we don't want to check them
                 continue;
             }
-            if(vj->is_adapted){
+            if (vj->realtime_level == nt::RTLevel::Adapted) {
```
Comment (TeXitoi) [inline source/kraken/tests/fill_disruption_from_chaos_tests.cpp]: same here

File: source/kraken/tests/fill_disruption_from_chaos_tests.cpp [hunk 3]
```diff
@@ -296,14 +296,14 @@ BOOST_AUTO_TEST_CASE(add_impact_and_update_on_stop_area) {
     BOOST_REQUIRE_EQUAL(b.data->pt_data->lines.size(), 1);
     BOOST_CHECK_EQUAL(b.data->pt_data->vehicle_journeys.size(), 2);
     BOOST_CHECK_EQUAL(b.data->pt_data->journey_patterns.size(), 1);
-    for(const auto* vj: b.data->pt_data->vehicle_journeys){
-        if(vj->is_adapted){
+    for (const auto* vj: b.data->pt_data->vehicle_journeys) {
+        if (vj->realtime_level == nt::RTLevel::Adapted) {
```
Comment (TeXitoi) [inline source/kraken/tests/fill_disruption_from_chaos_tests.cpp]: same here

File: source/routing/raptor.cpp [hunk 4]
```diff
@@ -410,13 +410,15 @@ void RAPTOR::set_valid_jp_and_jpp(
     uint32_t date,
     const type::AccessibiliteParams& accessibilite_params,
     const std::vector<std::string>& forbidden,
-    bool disruption_active)
+    const nt::RTLevel rt_level)
 {
 
-    if(disruption_active){
+    if (rt_level == nt::RTLevel::Adapted) {
```
Comment (TeXitoi) [inline source/routing/raptor.cpp]: I prefer a switch

File: source/routing/tests/raptor_test.cpp [hunk 5]
```diff
@@ -122,7 +122,7 @@ BOOST_AUTO_TEST_CASE(change){
     RAPTOR raptor(*(b.data));
     type::PT_Data & d = *b.data->pt_data;
 
-    auto res1 = raptor.compute(d.stop_areas[0], d.stop_areas[4], 7900, 0, DateTimeUtils::inf, false, true);
+    auto res1 = raptor.compute(d.stop_areas[0], d.stop_areas[4], 7900, 0, DateTimeUtils::inf, type::RTLevel::Theoric, true);
```
Comment (TeXitoi) [inline source/routing/tests/raptor_test.cpp]: long line

File: source/routing/tests/raptor_test.cpp [hunk 6]
```diff
@@ -148,7 +148,7 @@ BOOST_AUTO_TEST_CASE(change){
     BOOST_CHECK_EQUAL(DateTimeUtils::date(to_datetime(res.items[0].arrival, *(b.data))), 0);
     BOOST_CHECK_EQUAL(DateTimeUtils::date(to_datetime(res.items[1].arrival, *(b.data))), 0);
 
-    res1 = raptor.compute(d.stop_areas[0], d.stop_areas[4], 7900, 0, DateTimeUtils::set(0,8500), false, true);
+    res1 = raptor.compute(d.stop_areas[0], d.stop_areas[4], 7900, 0, DateTimeUtils::set(0,8500), type::RTLevel::Theoric, true);
```
Comment (TeXitoi) [inline source/routing/tests/raptor_test.cpp]: long line

File: source/type/pb_converter.cpp [hunk 7]
```diff
@@ -617,7 +617,7 @@ void fill_pb_object(const nt::VehicleJourney* vj,
         com->set_type("standard");
     }
     vehicle_journey->set_odt_message(vj->odt_message);
-    vehicle_journey->set_is_adapted(vj->is_adapted);
+    vehicle_journey->set_is_adapted(vj->realtime_level == nt::RTLevel::Adapted);
```
Comment (TeXitoi) [inline source/type/pb_converter.cpp:620]: What about the output of that?

File: source/type/type.cpp [hunk 8]
```diff
@@ -133,18 +133,22 @@ bool HasMessages::has_publishable_message(const boost::posix_time::ptime& curren
...
-bool StopTime::is_valid_day(u_int32_t day, const bool is_arrival, const bool is_adapted) const{
+bool StopTime::is_valid_day(u_int32_t day, const bool is_arrival, const RTLevel rt_level) const {
     if((is_arrival && arrival_time >= DateTimeUtils::SECONDS_PER_DAY)
        || (!is_arrival && departure_time >= DateTimeUtils::SECONDS_PER_DAY)) {
         if(day == 0)
             return false;
         --day;
     }
-    if(!is_adapted) {
+    if (rt_level == RTLevel::Theoric) {
```
Comment (TeXitoi) [inline source/type/type.cpp]: I prefer a switch

File: source/type/type.cpp [hunk 9]
```diff
@@ -187,14 +191,19 @@ uint32_t StopTime::f_departure_time(const u_int32_t hour, bool clockwise) const
     }
 }
 
-bool FrequencyVehicleJourney::is_valid(int day, const bool is_adapted) const {
+bool FrequencyVehicleJourney::is_valid(int day, const RTLevel rt_level) const {
     if (day < 0)
         return false;
-    if (! is_adapted) {
+    if (rt_level == RTLevel::Theoric) {
```
Comment (TeXitoi) [inline source/type/type.cpp]: I prefer a switch

File: source/type/type.h [hunk 10]
```diff
@@ -58,7 +59,9 @@ namespace navitia { namespace georef {
  struct Admin;
  struct GeoRef;
 }}
-namespace navitia { namespace type {
+namespace navitia {
+
+namespace type {
```
Comment (TeXitoi) [inline source/type/type.h]: Why?

File: source/type/type.h [hunk 11]
```diff
@@ -1134,5 +1137,6 @@ struct enum_size_trait<type::Mode_e> {
     }
 };
 
+
```
Comment (TeXitoi) [inline source/type/type.h]: Why?

File: source/type/pb_converter.cpp (hunk 7 above)
Comment (antoine-de) [inline source/type/pb_converter.cpp:620]: we'll need to change the API, but I think it has to be done later

**Comment by TeXitoi** [comment]
OK for me.

File: source/type/rt_level.h [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+namespace type {
+enum class RTLevel : char {
+    Theoric = 0,
+    Adapted,
+    RealTime
+};
+}
+
+template <>
+struct enum_size_trait<type::RTLevel> {
```
Comment (xlqian) [inline source/type/rt_level.h:45]: hmm, shouldn't 'flat_enum_map.h' be included in this file?

File: source/kraken/tests/fill_disruption_from_chaos_tests.cpp [hunk 13]
```diff
@@ -240,22 +246,28 @@ BOOST_AUTO_TEST_CASE(add_impact_and_update_on_stop_area) {
...
                 has_adapted_vj = true;
                 BOOST_CHECK_MESSAGE(boost::find_if(vj->journey_pattern->journey_pattern_point_list,
                             stop_area_finder("stop_area:stop1")) == vj->journey_pattern->journey_pattern_point_list.end(), dump_vj(*vj));
                 BOOST_CHECK_MESSAGE(boost::find_if(vj->journey_pattern->journey_pattern_point_list,
                             stop_area_finder("stop_area:stop2")) == vj->journey_pattern->journey_pattern_point_list.end(), dump_vj(*vj));
-            }else{
-                BOOST_CHECK(boost::find_if(vj->journey_pattern->journey_pattern_point_list,
-                            stop_area_finder("stop_area:stop1")) != vj->journey_pattern->journey_pattern_point_list.end());
+                break;
+            default:
```
Comment (pbougue) [inline source/kraken/tests/fill_disruption_from_chaos_tests.cpp:268]: Maybe add case nt::RTLevel::RealTime above to explicit TODO

File: source/type/rt_level.h (hunk 12 above)
Comment (antoine-de) [inline source/type/rt_level.h:45]: yup, done

Comment (xlqian) [inline source/type/rt_level.h:45]: ok

**Comment by xlqian** [comment]
ok for me

**Comment by pbougue** [comment]
OK for me

**Pull request merged**